File: src/main/java/dev/akif/cookbook/api/RecipeDTO.java
package dev.akif.cookbook.api;

import dev.akif.cookbook.recipe.Recipe;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public record RecipeDTO(
        long id,
        String name,
        int numberOfServings,
        Map<String, List<MeasuredIngredientDTO>> ingredients,
        List<String> instructions
) {
    public static RecipeDTO from(Recipe recipe) {
        return new RecipeDTO(
                recipe.id(),
                recipe.name(),
                recipe.numberOfServings(),
                recipe
                        .ingredients()
                        .entrySet()
                        .stream()
                        .map(e -> Map.entry(e.getKey(), e.getValue().stream().map(MeasuredIngredientDTO::from).toList()))
                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)),
                recipe.instructions()
        );
    }
}

File: src/main/kotlin/dev/akif/cookbook/api/RecipeDTO.kt
package dev.akif.cookbook.api

import dev.akif.cookbook.recipe.Recipe

data class RecipeDTO(
    val id: Long,
    val name: String,
    val numberOfServings: Int,
    val ingredients: Map<String, List<MeasuredIngredientDTO>>,
    val instructions: List<String>
) {
    companion object {
        fun from(recipe: Recipe): RecipeDTO =
            RecipeDTO(
                id = recipe.id,
                name = recipe.name,
                numberOfServings = recipe.numberOfServings,
                ingredients = recipe.ingredients.mapValues { (_, mi) -> mi.map(MeasuredIngredientDTO::from) },
                instructions = recipe.instructions,
            )
    }
}
